File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 3/섬 연결하기/섬 연결하기.py
def solution(n, costs):
    parents = [i for i in range(n)]
    costs = sorted(costs, key=lambda x:x[2])

    min_cost = 0
    for a,b,c in costs:
        if parents[a] == parents[b]:
            continue

        min_cost += c

        tmp = parents[b]
        for i in range(n):
            if parents[i] == tmp:
                parents[i] = parents[a]

        flag = True
        for i in range(1, n):
            if parents[i] != parents[0]:
                flag = False
                break

        if flag: break

    return min_cost

File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 3/섬 연결하기/섬 연결하기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool compare(vector<int> v1, vector<int> v2){
    return v1[2] < v2[2];
}

int solution(int n, vector<vector<int>> costs) {
    vector<int> parents;
    for(int i=0; i<n; parents.push_back(i++));
    
    // 비용이 적은 순으로 정렬
    sort(costs.begin(), costs.end(), compare);
    
    int min_cost=0;
    for(vector<int> cost : costs){
        // 이미 연결된 다리면 패스
        if(parents[cost[0]] == parents[cost[1]])
            continue;
        
        // 다리 연결하기
        min_cost += cost[2];
        
        int tmp = parents[cost[1]];
        int p = parents[cost[0]];
        for(int i=0; i<n; i++)
            if(parents[i] == tmp) parents[i] = p;
        
        // 모든 부모 노드가 같으면 끝내기
        bool flag = true;
        for(int i=1; i<n; i++){
            if(parents[i] != parents[0]){
                flag = false;
                break;
            }
        }
        
        if(flag) break;
    }
    
    return min_cost;
}
